source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
-        if (temp_sector > temp_int) {
-            heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
-            temp_sector = 0;
-        }
+    } else if (alloc_size > (heap_size - 2 * sizeof(int)) ) {
+        heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
+    } else if (alloc_size) {
+        block_count = (alloc_size + sizeof(int) - 1) / sizeof(int);
     }
-    return temp_sector;
+    return block_count;
 }
 
-static int8_t ns_sector_validate(int *sector_start, int direction)
+// Checks that block length indicators equals
+// Block has format: Size of data area in words | data area in words | Size of data area in words
+// If Size is negative it means area is unallocated
+// For direction, use 1 for direction up and -1 for down
+static int8_t ns_block_validate(int *block_start, int direction)
 {
     int8_t ret_val = -1;
-    int size_start, size_end;
-    if (direction) {
-        //Up
-        size_start = *sector_start++;
-        if (size_start < 0) {
-            sector_start += -(size_start);
-        } else {
-            sector_start += size_start;
-        }
-        size_end = *sector_start;
+    int *end = block_start;
+    int size_start = *end;
+    if (direction > 0) {
+        end += (1 + abs(size_start));
     } else {
-        size_end = *sector_start--;
-        if (size_end < 0) {
-            sector_start -= -(size_end);
-        } else {
-            sector_start -= size_end;
-        }
-        size_start = *sector_start;
+        end -= (1 + abs(size_start));
     }
-    if (size_start == size_end) {
+
+    if (size_start && size_start == *end) {
         ret_val = 0;
     }
     return ret_val;
 }
 #endif
 
-void *ns_dyn_mem_alloc(int16_t alloc_size)
+#ifdef STANDARD_MALLOC
+#ifdef USE_IAR
+#pragma optimize=none
+#endif
+#endif
+// For direction, use 1 for direction up and -1 for down
+static void *ns_dyn_mem_internal_alloc(const int16_t alloc_size, int direction)
 {
 #ifndef STANDARD_MALLOC
-
-    int *ptr = heap_main_end;
     void *retval = 0;
-    int h_size, alloc_sector, s_size;
-    int moved = 0;
-
-    alloc_sector = heap_alloc_internal_check(alloc_size);
-    if (alloc_sector) {
-        h_size = (heap_size / sizeof(int));
+    int alloc_block = heap_alloc_internal_check(alloc_size);
+    if (alloc_block) {
+        int h_size = (heap_size / sizeof(int));
+        int *ptr = direction > 0 ? heap_main : heap_main_end;
+        int moved = 0;
         platform_enter_critical();
-
         while (moved < h_size) {
-            if (ns_sector_validate(ptr, 0) == 0) {
-                s_size = *ptr;
-                if (s_size == 0) {
-                    heap_failure(NS_DYN_MEM_HEAP_SECTOR_CORRUPTED);
-                    retval = 0;
-                    break;
-                } else if (s_size < 0) {
-                    //Free
-                    if (((-s_size) >= alloc_sector)) {
+            if (ns_block_validate(ptr, direction) == 0) {
+                int block_total_size = *ptr;
```

Comment:
block_data_size. It's NOT the total size, which would include that two end markers.